### ISSUE TITLE
Peformance question for to_arrow, to_pandas, to_duckdb

### ISSUE THREAD

**Issue by jkleinkauff** [apache/iceberg-python#1032]
### Question

Hey, thanks for this very convenient library.

This is not a bug, just want to better understand something.

I have a question regarding the performance - ie time to query the table (?) - for such methods.

```python
if __name__ == "__main__":
    catalog = SqlCatalog(
        "default",
        **{
            "uri": f"postgresql+psycopg2://postgres:Password1@localhost/postgres",
        },
    )
    table = catalog.load_table("bronze.curitiba_starts_june")
    df = table.scan(limit=100)
    pa_table = df.to_arrow()
````
The code above will run ok. My question is regarding the last command, to_arrow() transformation takes around 60s (+-) to execute. I believe this is mostly because of the network itself? 
The execution time will stay roughly the same with different row limit (1, 10, 100).

Querying the same table in motherduck - using iceberg_scan - is faster:
<img width="709" alt="image" src="https://github.com/user-attachments/assets/98874655-4da3-4dcd-8110-767d19286f36">

When running the same query locally - without motherduck but still using duckdb - the execution time will be similar to what pyiceberg takes, actually it will be a bit slower. That's why I think this is mostly like a network "issue". Can you help me understand what's happening? Thank you!

#### Table Data
The table has two parquet files (110mb, 127mb)

**Comment by sungwy** [comment]
Hi @jkleinkauff , that's indeed an interesting observation.

I have some follow up questions to help us understand it better.
1. Where are your files stored?
2. Is there a way we can profile your IO and plot it against your IO Download limit?

As a way of comparison, I just ran a scan using to_arrow against a table that has 63, 5.5Mb Parquet files comprising the table. I'd imagine a table with less files to take less time to return (although the limit function here should ensure that we aren't even reading the other parquet files past the first one)

It returned in 6 seconds.


Your observation that limit 1~100 took similar times makes sense to me as well. If you have 100+ Mb files, you are going to have to download the same amount of data regardless to return the limited result.

**Comment by kevinjqliu** [comment]
There's a nontrivial cost in reading metadata files in Iceberg. 
Can you run this command,
```
table.inspect.manifests().to_pandas()
```
This will show the number of manifest files https://py.iceberg.apache.org/api/#manifests

**Comment by jkleinkauff** [comment]
Hey, thank you for taking a time to answer me!

1. My files are in S3.
2. Sure! It's something I could do on my end? Do you have any recommendation on that?
(I'm not sure if it's the same, running a download profiler (I've written something using psutil) in on file it takes something between 25s to complete it)

Yeah, even with limit=1 it seems scan is returning both files (just an observation, maybe it's intended):

```python
    df = table.scan(limit=1)
    # pa_table = df.to_arrow()
    [print(task.file.file_path) for task in df.plan_files()]
# s3://xxx/xxx/curitiba_starts_june/data/00000-0-6984da88-fe64-4765-9137-739072becfb1.parquet
# s3://xxx/xxx/curitiba_starts_june/data/00000-0-1de29b8f-2e8c-4543-9663-f769d53b17b7.parquet 

```

Output of table.inspect.manifests().to_pandas()

```python
❯ python pyiceberg_duckdb.py
   content                                               path  length  partition_spec_id  ...  added_delete_files_count  existing_delete_files_count  deleted_delete_files_count  partition_summaries
0        0  s3://data-lake-jho/bronze/curitiba_starts_june...   10433                  0  ...                         0                            0                           0                   []
1        0  s3://data-lake-jho/bronze/curitiba_starts_june...   10430                  0  ...                         0                            0                           0                   []

[2 rows x 12 columns]
```
I can also share the files or a direct link to my files. Thank you!

**Comment by jkleinkauff** [comment]
Hi @kevinjqliu thank you for your time!

Those are my findings:

I've included a read_parquet method from awswrangler. Don't know why, but it's by far the fastest method.

reading raw parquet file with awswrangler:
[read_raw_parquet_awswrangler.py](https://gist.github.com/jkleinkauff/acf9e876a72201c9530a0070dbb25e6e)
```
1st run: Time taken to read the Parquet file: 16.10819697380066 seconds
2nd run: Time taken to read the Parquet file: 16.06696915626526 seconds
3rd run: Time taken to read the Parquet file: 14.28455901145935 seconds
```

reading raw parquet file with pyarrow
[read_raw_parquet_pyarrow.py](https://gist.github.com/jkleinkauff/a31538c06ee1d4622a03cbf98a369ce8)
```
1st run: Time taken to read the Parquet file: 39.86264896392822 seconds
2nd run: Time taken to read the Parquet file: 39.484612226486206 seconds
3rd run: Time taken to read the Parquet file: 26.693129062652588 seconds
```

reading entire iceberg table, without limits
[read_iceberg_full.py](https://gist.github.com/jkleinkauff/f828c9a282180417db94dbc98497691f)
```
1st run: Time taken to read the Iceberg table: 21.632921934127808 seconds
2nd run: Time taken to read the Iceberg table: 36.94430899620056 seconds
3rd run: Time taken to read the Iceberg table: 49.66138482093811 seconds
```

reading iceberg table, with limit of 1
[read_iceberg_limit.py](https://gist.github.com/jkleinkauff/2e92add21bf84f18b5cb66d43aa7afe0)
```
1st run: Time taken to read the Iceberg table: 45.886711835861206 seconds
2nd run: Time taken to read the Iceberg table: 29.464744091033936 seconds
3rd run: Time taken to read the Iceberg table: 44.78428387641907 seconds
```

reading iceberg table with duckdb
[read_iceberg_duckdb.py](https://gist.github.com/jkleinkauff/9a48db4383516c13b6403c30468bd856)
```
1st run: Time taken to read the Parquet file: 59.5912652015686 seconds
2nd run: Time taken to read the Parquet file: 61.646626710891724 seconds
3rd run: Time taken to read the Parquet file: 58.97534728050232 seconds

Proxying through motherduck (con = duckdb.connect("md:db")):
1st run: Time taken to read the Parquet file: 105.63072204589844 seconds
2nd run: Time taken to read the Parquet file: 144.91437602043152 seconds
3rd run: Time taken to read the Parquet file: 176.27135396003723 seconds
```

eading iceberg table with duckdb, with limit of 1
[read_iceberg_duckdb_limit.py](https://gist.github.com/jkleinkauff/04883f509a8df8a27c6c808bede0b6c1)
```
1st run: Time taken to read the Parquet file: 63.78661298751831 seconds
2nd run: Time taken to read the Parquet file: 79.1733546257019 seconds
3rd run: Time taken to read the Parquet file: 80.755441904068 seconds
Proxying through motherduck:
Why md is faster here? It somehow is pushing it
1st run: Time taken to read the Parquet file: 3.524472951889038 seconds
2nd run: Time taken to read the Parquet file: 3.4903008937835693 seconds
3rd run: Time taken to read the Parquet file: 3.258246898651123 seconds
```

**Comment by kevinjqliu** [comment]
Thanks for looking into the different scenarios. It looks like there are varying results depending on the engines.

### Read Path
I took a deeper look into the read path for PyIceberg and discovered a missing optimization. 
For context, reading an Iceberg table is done through the internal TableScan API. For example, `catalog.load_table(“foo.bar”).scan().to_arrow()` produces a pyarrow table. Tracing the call stack, here are the main areas to look at
* [`to_arrow`](https://github.com/apache/iceberg-python/blob/d8b5c17cadbc99e53d08ade6109283ee73f0d83e/pyiceberg/table/__init__.py#L2033-L2044), gathers the TableScan parameters and materializes a Pyarrow Table. It calls `project_table` with an iterable of `FileScanTask` (`self.plan_files()`) and specified limit, if any.
* [`project_table`](https://github.com/apache/iceberg-python/blob/d8b5c17cadbc99e53d08ade6109283ee73f0d83e/pyiceberg/io/pyarrow.py#L1361-L1386), takes the FileScanTasks, which corresponds to parquet files, and processes them asynchronously with futures. The futures will stop early if the limit is satisfied.

So given an Iceberg table with 2 files. `scan(limit=1).to_arrow()` will produce 2 FileScanTasks (corresponding with each parquet file). Both tasks will be submitted simultaneously. After reading the entire parquet file, one of the tasks is completed. The limit is satisfied and the other future is cancelled.

The important note here is that the unit of work is per FileScanTask, which means per parquet file read. The limit will be checked only after each future completion. This means at least 1 parquet file will be read entirely before the limit is applied. The limit is never pushed down to the parquet file read level. **Even when `limit=1` is specified, the entire parquet file will be read**. In your example, with or without limit, an entire ~100mb data file will be read. 

The potential optimization here is to pushdown the `limit` to [`_task_to_table`](https://github.com/apache/iceberg-python/blob/d8b5c17cadbc99e53d08ade6109283ee73f0d83e/pyiceberg/io/pyarrow.py#L1262) and in turn to [`_task_to_record_batches`](https://github.com/apache/iceberg-python/blob/d8b5c17cadbc99e53d08ade6109283ee73f0d83e/pyiceberg/io/pyarrow.py#L1191). Perhaps check the limit against the number of records [when yielding each batch](https://github.com/apache/iceberg-python/blob/d8b5c17cadbc99e53d08ade6109283ee73f0d83e/pyiceberg/io/pyarrow.py#L1240-L1259).
In fact, this optimization is already done in [`project_batches` ](https://github.com/apache/iceberg-python/blob/d8b5c17cadbc99e53d08ade6109283ee73f0d83e/pyiceberg/io/pyarrow.py#L1457-L1475), which is used by [`to_arrow_batch_reader`](https://github.com/apache/iceberg-python/blob/d8b5c17cadbc99e53d08ade6109283ee73f0d83e/pyiceberg/table/__init__.py#L2046-L2063) (an alternative to `to_arrow`). 

### Unify implementations
As a side note, there might be an opportunity to unify the implementation details of `to_arrow` and `to_arrow_batch_reader`.

### `to_arrow_batch_reader` bug
There is also a bug with `to_arrow_batch_reader`, as it does not respecting the given `limit`, returning more records than specified. The bug is in [`project_batches`](https://github.com/apache/iceberg-python/blob/d8b5c17cadbc99e53d08ade6109283ee73f0d83e/pyiceberg/io/pyarrow.py#L1457-L1475), specifically with the way yield interacts with the two for-loops. Here’s a [Jupyter notebook](https://gist.github.com/kevinjqliu/1b5bc445ed12ca78e564db139c1e5683) reproducing the issue, see the last cell and the number of rows read by using `to_arrow` vs `to_arrow_batch_reader`.

## Summary
* Optimize `to_arrow` to push down `limit` to the file level
* Potentially unify implementation of `to_arrow` and `to_arrow_batch_reader`
* Fix `to_arrow_batch_reader` to respect the given `limit`

**Comment by jkleinkauff** [comment]
@kevinjqliu that's awesome! Thank you so much !
I have one more question regarding the **read_parquet** from awswrangler. 
Do you know why it's faster than the other methods? Is there any optimization on their end or something?

**Comment by corleyma** [comment]
@jkleinkauff Without profiling it's hard to say for sure, but I can make a bet based on my experience with object storage clients/fsspec implementations in particular:
- you're benchmarking the fsspec FileIO path in pyiceberg, which if I understand correctly is using fsspec s3fs directly with a lot of defaults.  Probably it keeps the default block size (5mb).
- awswrangler is probably using a much bigger block size: https://github.com/aws/aws-sdk-pandas/blob/main/awswrangler/s3/_read_parquet.py#L170 (for a read like yours, block size is 20mb).

This is one thing that will have a big impact, but there are likely other differences too since, as you can see, awswrangler is not using the fsspec s3fs implementation and has its own wrappers for i/o over S3...  There's also a lot of stuff I see there in awswranglerthat is similar to some of the optimizations pyarrow makes in its native blob storage filesystem implementations;  maybe the fsspec s3fs implementation supports some of these too, but I'm less familiar with current state of that project.   At least historically, a lot of the defaults were tuned for random access, and not really the best for maximizing throughput in columnar workloads; pyiceberg probably should not be using all defaults.

I believe pyiceberg also provides FileIO implementations that use pyarrow native filesystems instead of fsspec.  I am curious what you find if you benchmark that? There may also be defaults there that pyiceberg should be tuning, but I suspect they will still be better tuned for the kinds of reads pyiceberg is doing.

**Comment by kevinjqliu** [comment]
> I have one more question regarding the read_parquet from awswrangler.
Do you know why it's faster than the other methods? Is there any optimization on their end or something?

I was also surprised by the performance difference. It's hard for me to say unless I look into the implementation details (in [awswrangler/s3/_read_parquet.py](https://github.com/aws/aws-sdk-pandas/blob/8d0c071649fb9e603a2ab2846307f902fafeabf5/awswrangler/s3/_read_parquet.py#L318)). There's definitely room for optimizations on the PyIceberg side. 

If you look at another engine like [daft](https://www.getdaft.io/projects/docs/en/latest/user_guide/basic_concepts/read-and-write.html#from-files), which is optimized for reading parquet on S3, that's a good target for potential performance gains. 

On the PyIceberg side, there's a future opportunity to integrate with iceberg-rust, which might speed up reading files.

**Comment by kevinjqliu** [comment]
> you're benchmarking the fsspec FileIO path in pyiceberg, which if I understand correctly is using fsspec s3fs directly with a lot of defaults. Probably it keeps the default block size (5mb).

There are two FileIO implementations, [fsspec](https://github.com/apache/iceberg-python/blob/f05b1aedee8451d981188adf68be5e8b360a9ca1/pyiceberg/io/fsspec.py#L250) and [pyarrow](https://github.com/apache/iceberg-python/blob/f05b1aedee8451d981188adf68be5e8b360a9ca1/pyiceberg/io/pyarrow.py). In the case above, I believe pyarrow is used, since its preferred over fsspec ([source](https://github.com/apache/iceberg-python/blob/f05b1aedee8451d981188adf68be5e8b360a9ca1/pyiceberg/io/__init__.py#L289-L291))

Looks like the pyarrow default buffer size is 1MB
https://github.com/apache/iceberg-python/blob/f05b1aedee8451d981188adf68be5e8b360a9ca1/pyiceberg/io/pyarrow.py#L417

**Comment by jkleinkauff** [comment]
@sungwy @kevinjqliu I'm really enjoying this discussion and learning a ton from it. Would love to keep it going but feel free to close it as this is not an issue. Thank you folks!

**Comment by kevinjqliu** [comment]
Thanks for reporting this. I learned a lot from exploring this thread, and we have some solid improvements coming up. Please let us know if anything else comes up!
